locations/spiders/toyota_eu.py

```diff
@@ -1,65 +1,86 @@
+from locations.pipelines.address_clean_up import clean_address
+from locations.spiders.toyota_au import TOYOTA_SHARED_ATTRIBUTES
 
 
-class ToyotaEUSpider(scrapy.Spider):
+class ToyotaEUSpider(JSONBlobSpider):
     name = "toyota_eu"
-    item_attributes = {
-        "brand": "Toyota",
-        "brand_wikidata": "Q53268",
-    }
-    available_countries = [
-        "tr",
-        "az",
-        "pl",
-        "ru",
-        "lv",
-        "si",
-        "lt",
-        "ua",
-        "pt",
-        "se",
-        "ba",
-        "is",
-        "sk",
-        "de",
-        "no",
-        "ro",
-        "es",
-        "am",
-        "hr",
-        "hu",
-        "fi",
-        "ge",
-        "fr",
-        "nl",
-        "gr",
-        "it",
-        "rs",
-        "bg",
-        "cz",
+    item_attributes = TOYOTA_SHARED_ATTRIBUTES
+    locations_key = "dealers"
+
+    all_countries = [country.alpha_2.lower() for country in pycountry.countries]
+    exclude_countries = [
+        "bf",  # toyota_africa
+        "bj",
+        "cd",
+        "cf",
+        "cg",
+        "ci",
+        "cm",
+        "gm",
+        "gn",
+        "gq",
+        "gw",
+        "ke",
+        "lr",
+        "mg",
+        "ml",
+        "mr",
+        "mz",
+        "ne",
+        "ng",
+        "rw",
+        "sl",
+        "sn",
+        "td",
+        "tg",
+        "ug",
+        "zw",
+        "au",  # toyota_au
+        "br",  # toyota_br
+        "nz",  # toyota_nz
+        "bw",  # toyota_sacu
+        "ls",
+        "na",
+        "sz",
+        "za",
+        "us",  # toyota_us
     ]
 
     def start_requests(self):
-        for country in self.available_countries:
+        available_countries = [c for c in self.all_countries if c not in self.exclude_countries]
+        for country in available_countries:
             yield scrapy.Request(
-                f"https://kong-proxy-aws.toyota-europe.com/dxp/dealers/api/toyota/{country}/cs/all?extraCountries=&services=&randomize=false",
+                f"https://kong-proxy-intranet.toyota-europe.com/dxp/dealers/api/toyota/{country}/en/all",
                 callback=self.parse,
             )
 
-    def parse(self, response):
-        for store in response.json().get("dealers"):
-            address_details = store["address"]
-            coordinates = address_details["geo"]
-            item = DictParser.parse(store)
-            item["ref"] = store["id"]
-            item["email"] = store["eMail"]
-            item["lat"] = coordinates["lat"]
-            item["lon"] = coordinates["lon"]
-            item["country"] = store["country"].upper()
+    def pre_process_data(self, feature):
+        feature["geo"] = feature["address"].pop("geo", None)
+        feature["email"] = feature.pop("eMail")
 
-            apply_category(Categories.SHOP_CAR, item)
+    def post_process_item(self, item, response, location):
+        apply_category(Categories.SHOP_CAR, item)
```

Comment:
May I kindly ask to improve categories a bit? There are usually not only car shops in the data, but also repairs and car parts shops. Should be something like in [Citroen spider](https://github.com/alltheplaces/alltheplaces/blob/41ef4db4225bef8e67b0870d1eedc55c5986bd3c/locations/spiders/citroen.py#L62-L70).

Reply:
Good point, there is data on services offered, so I will have a look and map that to categories